lib/fluent/plugin/kinesis_helper/api.rb

```diff
@@ -164,6 +177,49 @@ def give_up_retries(failed_records)
                 record[:original]
               ])
             }
+
+            if @raise_error_on_batch_request_failure
+              # Raise error and return chunk to Fluentd for retrying
+              case request_type
+              # @see https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecords.html
+              # @see https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Kinesis/Client.html#put_records-instance_method
+              # @see https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Kinesis/Errors.html
+              when :streams, :streams_aggregated
+                provisioned_throughput_exceeded_records = failed_records.select { |record| record[:error_code] == 'ProvisionedThroughputExceededException' }
+                target_failed_record = provisioned_throughput_exceeded_records.first || failed_records.first
+                target_error = provisioned_throughput_exceeded_records.empty? ?
+                  Aws::Kinesis::Errors::ServiceError :
+                  Aws::Kinesis::Errors::ProvisionedThroughputExceededException
+              # @see https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecords.html
+              # @see https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Firehose/Client.html#put_record_batch-instance_method
+              # @see https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Firehose/Errors.html
+              when :firehose
+                service_unavailable_exception_records = failed_records.select { |record| record[:error_code] == 'ServiceUnavailableException' }
+                target_failed_record = service_unavailable_exception_records.first || failed_records.first
+                target_error = service_unavailable_exception_records.empty? ?
+                  Aws::Firehose::Errors::ServiceError :
+                  Aws::Firehose::Errors::ServiceUnavailableException
+              end
+              log.error("Raise #{target_failed_record[:error_code]} and return chunk to Fluentd buffer for retrying")
+              raise target_error.new(Seahorse::Client::RequestContext.new, target_failed_record[:error_message])
+            else
+              # Increment num_errors to monitor batch request failure from "monitor_agent" or "fluent-plugin-prometheus"
+              increment_num_errors
```

Comment:
I may not understand this intention accurately, but I have some questions.

1. Is it ok to call `increment_num_errors` regardless of `@monitor_num_of_batch_request_retries`?
2. Is it ok not to call `increment_num_errors` in the `if-true` case of L181?

It's a bit hard to follow when `increment_num_errors` is called and when it's not.

It would be straightforward if we can jut add the following line at L96 (in the new commit) instead of everything, but I'm not sure if it's appropriate.
```ruby
increment_num_errors if @monitor_num_of_batch_request_retries
```

Reply:
Thanks for your comments! Yes, it would work as I expected. This enhancement will work as follows:

### current behavior
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* No metrics will be updated when giving up request retries (= CANNOT monitor anything from monitor_agent)

### implemented behavior: default
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by this plugin itself when giving up request retries (= can monitor from monitor_agent)

### implemented behavior: raise_error_on_batch_request_failure = true
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will return records to Fluentd buffer (= will retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by Fluentd retry mechanism when giving up request retries (= can monitor from monitor_agent)

### implemented behavior: monitor_num_of_batch_request_retries = true
* *num_errors* metrics will be updated by batch request retries due to Kinesis over capacity or any other causes (= can monitor from monitor_agent)
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by this plugin itself when giving up request retries (= can monitor from monitor_agent)

Reply:
Thanks for explaining! Now I understand the expected behavior and I think the current implementation will work as expected. I didn't know fluentd will retry when error raised, and increase `@num_errors`.

If I add one more thing, the config name, `raise_error_on_batch_request_failure`, is about the implementation detail. It might be better if it explains higher-level behavior (e.g. `ignore_batch_request_failure_after_retry` with default value: true).

Reply:
Thank you for your feedback @jooohn, I totally agree with you about config name. I've renamed `raise_error_on_batch_request_failure` to `drop_failed_records_after_batch_request_retries` with default true.

What do you think about it?

Reply:
LGTM!